File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/DataStruct/NoOrPriorityList/PriorityList/PriorityList.cpp
#include "MyProject.h"
#include "PriorityList.h"
#include "MacroDef.h"
#include "Ctx.h"
#include "LogSys.h"
#include "LogTypeId.h"
#include "PrioritySort.h"
#include "MyMemoryConstructorFlag.h"
#include "MyMemoryAllocatorConfig.h"
#include "MyMemoryDefaultAlloc.h"
#include "PriorityProcessObject.h"

MY_BEGIN_NAMESPACE(MyNS)

PriorityList::PriorityList()
{
	//this->mPriorityProcessObjectList = new MList<PriorityProcessObject>();
	this->mPrioritySort = PrioritySort::ePS_Great;
	this->mIsSpeedUpFind = false;
	this->mIsOpKeepSort = false;
}

void PriorityList::init()
{

}

void PriorityList::dispose()
{
	//if(nullptr != this->mPriorityProcessObjectList)
	//{
		this->mPriorityProcessObjectList.dispose();
		//this->mPriorityProcessObjectList = nullptr;
	//}
	//if(nullptr != this->mDic)
	//{
		this->mDic.dispose();
		//this->mDic = nullptr;
	//}
}

void PriorityList::setIsSpeedUpFind(bool value)
{
	this->mIsSpeedUpFind = value;

	//if (this->mIsSpeedUpFind)
	//{
	//	this->mDic = new MDictionary<INoOrPriorityObject, int>();
	//}
}

void PriorityList::setIsOpKeepSort(bool value)
{
	this->mIsOpKeepSort = value;
}

void PriorityList::clear()
{
	int index = 0;
	int listLen = this->count();

	while(index < listLen)
	{
		//NumIdBufferObjectFactory.deleteObject(this->mPriorityProcessObjectList.get(index));
		MY_DELETE this->mPriorityProcessObjectList.get(index);
		this->mPriorityProcessObjectList.set(index, nullptr);
		index += 1;
	}

	this->mPriorityProcessObjectList.clear();

	if(this->mIsSpeedUpFind)
	{
		this->mDic.clear();
	}
}

int PriorityList::count()
{
	return this->mPriorityProcessObjectList.count();
}

INoOrPriorityObject* PriorityList::get(int index)
{
	INoOrPriorityObject* ret = nullptr;

	if(index < this->count())
	{
		ret = this->mPriorityProcessObjectList.get(index)->mPriorityObject;
	}

	return ret;
}

float PriorityList::getPriority(int index)
{
	float ret = 0;

	if (index < this->count())
	{
		ret = this->mPriorityProcessObjectList.get(index)->mPriority;
	}

	return ret;
}

bool PriorityList::contains(INoOrPriorityObject* item)
{
	bool ret = false;

	if (nullptr != item)
	{
		if (this->mIsSpeedUpFind)
		{
			ret = this->mDic.containsKey(item);
		}
		else
		{
			int index = 0;
			int listLen = this->mPriorityProcessObjectList.count();

			while (index < listLen)
			{
				if (item == this->mPriorityProcessObjectList.get(index)->mPriorityObject)
				{
					ret = true;
					break;
				}

				++index;
			}
		}
	}
	else
	{
		if (MacroDef::ENABLE_LOG)
		{
			GLogSys->log("PriorityList::Contains, failed", LogTypeId::eLogPriorityListCheck);
		}
	}

	return ret;
}

void PriorityList::removeAt(int index)
{
	PriorityProcessObject* priorityProcessObject = nullptr;
	priorityProcessObject = this->mPriorityProcessObjectList.get(index);

	if (this->mIsSpeedUpFind)
	{
		this->_effectiveRemove(this->mPriorityProcessObjectList.get(index)->mPriorityObject);
	}
	else
	{
		this->mPriorityProcessObjectList.removeAt(index);
	}

	if (nullptr != priorityProcessObject)
	{
		//NumIdBufferObjectFactory.deleteObject(priorityProcessObject);
		MY_DELETE priorityProcessObject;
		priorityProcessObject = nullptr;
	}
}

int PriorityList::getIndexByPriority(float priority)
{
	int retIndex = -1;

	int index = 0;
	int listLen = this->mPriorityProcessObjectList.count();

	while (index < listLen)
	{
		if (PrioritySort::ePS_Less == this->mPrioritySort)
		{
			if (this->mPriorityProcessObjectList.get(index)->mPriority >= priority)
			{
				retIndex = index;
				break;
			}
		}
		else if (PrioritySort::ePS_Great == this->mPrioritySort)
		{
			if (this->mPriorityProcessObjectList.get(index)->mPriority <= priority)
			{
				retIndex = index;
				break;
			}
		}

		++index;
	}

	return retIndex;
}

int PriorityList::getIndexByPriorityObject(INoOrPriorityObject* priorityObject)
{
	int retIndex = -1;

	int index = 0;
	int listLen = this->mPriorityProcessObjectList.count();

	while (index < listLen)
	{
		if (this->mPriorityProcessObjectList.get(index)->mPriorityObject == priorityObject)
		{
			retIndex = index;
			break;
		}

		++index;
	}

	return retIndex;
}

int PriorityList::getIndexByNoOrPriorityObject(INoOrPriorityObject* priorityObject)
{
	return this->getIndexByPriorityObject(priorityObject);
}

void PriorityList::addPriorityObject(INoOrPriorityObject* priorityObject, float priority)
{
	if (nullptr != priorityObject)
	{
		if (!this->contains(priorityObject))
		{
			PriorityProcessObject* priorityProcessObject = nullptr;
			priorityProcessObject = MY_NEW PriorityProcessObject();
			//priorityProcessObject = NumIdBufferObjectFactory.newObject<PriorityProcessObject>(BufferType.eBT_PriorityProcessObject, true);

			priorityProcessObject->mPriorityObject = priorityObject;
			priorityProcessObject->mPriority = priority;

			if (!this->mIsOpKeepSort)
			{
				this->mPriorityProcessObjectList.add(priorityProcessObject);

				if (this->mIsSpeedUpFind)
				{
					this->mDic.add(priorityObject, this->mPriorityProcessObjectList.count() - 1);
				}
			}
			else
			{
				int index = this->getIndexByPriority(priority);

				if (-1 == index)
				{
					this->mPriorityProcessObjectList.add(priorityProcessObject);

					if (this->mIsSpeedUpFind)
					{
						this->mDic.add(priorityObject, this->mPriorityProcessObjectList.count() - 1);
					}
				}
				else
				{
					this->mPriorityProcessObjectList.insert(index, priorityProcessObject);

					if (this->mIsSpeedUpFind)
					{
						this->mDic.add(priorityObject, index);
						this->_updateIndex(index + 1);
					}
				}
			}
		}
	}
	else
	{
		if (MacroDef::ENABLE_LOG)
		{
			GLogSys->log("PriorityList::addPriorityObject, failed", LogTypeId::eLogPriorityListCheck);
		}
	}
}

void PriorityList::removePriorityObject(INoOrPriorityObject* priorityObject)
{
	if (this->contains(priorityObject))
	{
		if (this->mIsSpeedUpFind)
		{
			this->_effectiveRemove(priorityObject);
		}
		else
		{
			int index = this->getIndexByPriorityObject(priorityObject);

			if(-1 != index)
			{
				this->mPriorityProcessObjectList.removeAt(index);
			}
		}
	}
}

void PriorityList::addNoOrPriorityObject(INoOrPriorityObject* noPriorityObject, float priority = 0.0f)
{
	this->addPriorityObject(noPriorityObject, priority);
}

void PriorityList::removeNoOrPriorityObject(INoOrPriorityObject* noPriorityObject)
{
	this->removePriorityObject(noPriorityObject);
}

// 快速移除元素
bool PriorityList::_effectiveRemove(INoOrPriorityObject* item)
{
	bool ret = false;

	if (this->mDic.containsKey(item))
	{
		ret = true;

		int index = this->mDic.value(item);
		this->mDic.remove(item);

		if (index == this->mPriorityProcessObjectList.count() - 1)    // 如果是最后一个元素，直接移除
		{
			this->mPriorityProcessObjectList.removeAt(index);
		}
		else
		{
			// 这样移除会使优先级顺序改变
			if (!this->mIsOpKeepSort)
			{
				this->mPriorityProcessObjectList.set(index, this->mPriorityProcessObjectList.get(this->mPriorityProcessObjectList.count() - 1));
				this->mPriorityProcessObjectList.removeAt(this->mPriorityProcessObjectList.count() - 1);
				this->mDic.add(this->mPriorityProcessObjectList.get(index)->mPriorityObject, index);
			}
			else
			{
				this->mPriorityProcessObjectList.removeAt(index);
				this->_updateIndex(index);
			}
		}
	}

	return ret;
}

void PriorityList::_updateIndex(int index)
{
	int listLen = this->mPriorityProcessObjectList.count();

	while (index < listLen)
	{
		this->mDic.add(this->mPriorityProcessObjectList.get(index)->mPriorityObject, index);

		++index;
	}
}

MY_END_NAMESPACE

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/DataStruct/NoOrPriorityList/PriorityList/PriorityList.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

MLoader("MyLua.Libs.DataStruct.MList");
MLoader("MyLua.Libs.DataStruct.MDictionary");
MLoader("MyLua.Libs.DataStruct.NoOrPriorityList.INoOrPriorityList");
MLoader("MyLua.Libs.DataStruct.NoOrPriorityList.PriorityList.PrioritySort");

-- 优先级队列
local M = GlobalNS.Class(GlobalNS.INoOrPriorityList);
M.clsName = "PriorityList";
GlobalNS[M.clsName] = M;

function M:ctor()
	self.mPriorityProcessObjectList = GlobalNS.new(GlobalNS.MList);
	self.mPrioritySort = GlobalNS.PrioritySort.ePS_Great;
	self.mIsSpeedUpFind = false;
	self.mIsOpKeepSort = false;
end

function M:setIsSpeedUpFind(value)
	self.mIsSpeedUpFind = value;

	if (self.mIsSpeedUpFind) then
		self.mDic = GlobalNS.new(GlobalNS.MDictionary);
	end
end

function M:setIsOpKeepSort(value)
	self.mIsOpKeepSort = value;
end

function M:clear()
	self.mPriorityProcessObjectList:clear();

	if(self.mIsSpeedUpFind) then
		self.mDic:clear();
	end
end

function M:count()
	return self.mPriorityProcessObjectList:count();
end

function M:get(index)
	local ret = nil;

	if(index < self:count()) then
		ret = self.mPriorityProcessObjectList:get(index).mPriorityObject;
	end

	return ret;
end

function M:getPriority(index)
	local ret = 0;

	if (index < self:count()) then
		ret = self.mPriorityProcessObjectList:get(index).mPriority;
	end

	return ret;
end

function M:contains(item)
	local ret = false;

	if (nil ~= item) then
		if (self.mIsSpeedUpFind) then
			ret = self.mDic:containsKey(item);
		else
			local index = 0;
			local listLen = self.mPriorityProcessObjectList:count();

			while (index < listLen) do
				if (item == self.mPriorityProcessObjectList:get(index).mPriorityObject) then
					ret = true;
					break;
				end

				index = index + 1;
			end
		end
	else
		if (MacroDef.ENABLE_LOG) then
			Ctx.mLogSys:log("PriorityList::Contains, failed", GlobalNS.LogTypeId.eLogPriorityListCheck);
		end
	end

	return ret;
end

function M:removeAt(index)
	if (self.mIsSpeedUpFind) then
		self:effectiveRemove(self.mPriorityProcessObjectList:at(index).mPriorityObject);
	else
		self.mPriorityProcessObjectList:removeAt(index);
	end
end

function M:getIndexByPriority(priority)
	local retIndex = -1;

	local index = 0;
	local listLen = self.mPriorityProcessObjectList:count();

	while (index < listLen) do
		if (GlobalNS.PrioritySort.ePS_Less == self.mPrioritySort) then
			if (self.mPriorityProcessObjectList:get(index).mPriority >= priority) then
				retIndex = index;
				break;
			end
		elseif (GlobalNS.PrioritySort.ePS_Great == self.mPrioritySort) then
			if (self.mPriorityProcessObjectList:get(index).mPriority <= priority) then
				retIndex = index;
				break;
			end
		end

		index = index + 1;
	end

	return retIndex;
end

function M:getIndexByPriorityObject(priorityObject)
	local retIndex = -1;

	local index = 0;
	local listLen = self.mPriorityProcessObjectList:count();

	while (index < listLen) do
		if (self.mPriorityProcessObjectList:get(index).mPriorityObject == priorityObject) then
			retIndex = index;
			break;
		end

		index = index + 1;
	end

	return retIndex;
end

function M:getIndexByNoOrPriorityObject(priorityObject)
	return self:getIndexByPriorityObject(priorityObject);
end

function M:addPriorityObject(priorityObject, priority)
	if (nil ~= priorityObject) then
		if (not self:contains(priorityObject)) then
			local priorityProcessObject = nil;
			priorityProcessObject = GlobalNS.new(GlobalNS.PriorityProcessObject);

			priorityProcessObject.mPriorityObject = priorityObject;
			priorityProcessObject.mPriority = priority;

			if (not self.mIsOpKeepSort) then
				self.mPriorityProcessObjectList:add(priorityProcessObject);

				if (self.mIsSpeedUpFind) then
					self.mDic:add(priorityObject, self.mPriorityProcessObjectList:count() - 1);
				end
			else
				local index = self:getIndexByPriority(priority);

				if (-1 == index) then
					self.mPriorityProcessObjectList:add(priorityProcessObject);

					if (self.mIsSpeedUpFind) then
						self.mDic:add(priorityObject, self.mPriorityProcessObjectList:count() - 1);
					end
				else
					self.mPriorityProcessObjectList:Insert(index, priorityProcessObject);

					if (self.mIsSpeedUpFind) then
						self.mDic:add(priorityObject, index);
						self:updateIndex(index + 1);
					end
				end
			end
		end
	else
		if (MacroDef.ENABLE_LOG) then
			Ctx.mLogSys:log("PriorityList::addPriorityObject, failed", GlobalNS.LogTypeId.eLogPriorityListCheck);
		end
	end
end

function M:removePriorityObject(priorityObject)
	if (self:contains(priorityObject)) then
		if (self.mIsSpeedUpFind) then
			self:effectiveRemove(priorityObject);
		else
			local index = self:getIndexByPriorityObject(priorityObject);

			if(-1 ~= index) then
				self.mPriorityProcessObjectList:removeAt(index);
			end
		end
	end
end

function M:addNoOrPriorityObject(noPriorityObject, priority)
	self:addPriorityObject(noPriorityObject);
end

function M:removeNoOrPriorityObject(noPriorityObject)
	self:removePriorityObject(noPriorityObject);
end

-- 快速移除元素
function M:effectiveRemove(item)
	local ret = false;

	if (self.mDic:containsKey(item)) then
		ret = true;

		local index = self.mDic:value(item);
		self.mDic:remove(item);

		if (index == self.mPriorityProcessObjectList:count() - 1) then    -- 如果是最后一个元素，直接移除
			self.mPriorityProcessObjectList:removeAt(index);
		else
			-- 这样移除会使优先级顺序改变
			if (not self.mIsOpKeepSort) then
				self.mPriorityProcessObjectList:set(index, self.mPriorityProcessObjectList:get(self.mPriorityProcessObjectList:count() - 1));
				self.mPriorityProcessObjectList:removeAt(self.mPriorityProcessObjectList:count() - 1);
				self.mDic:add(self.mPriorityProcessObjectList:get(index).mPriorityObject, index);
			else
				self.mPriorityProcessObjectList:removeAt(index);
				self:updateIndex(index);
			end
		end
	end

	return ret;
end

function M:updateIndex(index)
	local listLen = self.mPriorityProcessObjectList:count();

	while (index < listLen) do
		self.mDic:add(self.mPriorityProcessObjectList:get(index).mPriorityObject, index);

		index = index + 1;
	end
end

return M;